From 0000000000000000000000000000000000000030 Mon Sep 17 00:00:00 2001
From: steamac <steamac@localhost>
Date: Tue, 6 Oct 2026 18:02:00 +0300
Subject: [PATCH 30/30] MSL: Use normal function-call emission for mesh-stream helpers

STEAMAC-1B, STEAMAC-1D: the non-void geometry-helper regression also
exercises helpers that emit vertices, directly and through another helper.
Once their return types are preserved, the fork's separate OpFunctionCall
emitter leaves result temporaries inside a new scope and redeclares implicit
stage-input/output arguments as uninitialized locals instead of forwarding
them. It also omits normal call dependency/out-variable handling.

Delete the duplicated call emitter. Use the existing append_global_func_args
hook to append the mesh stream after the normal arguments, and match that
order in the helper prototype. Normal GLSL/MSL emission preserves result
scope, transitive global references, array returns and call dependencies.
---

diff --git a/spirv_msl.cpp b/spirv_msl.cpp
--- a/spirv_msl.cpp
+++ b/spirv_msl.cpp
@@ -11763,122 +11763,6 @@
 		break;
 	}
 
-	case OpFunctionCall:
-	{
-		uint32_t func_id = ops[2];
-		if (msl_options.for_mesh_pipeline && functions_needing_mesh_stream.count(func_id))
-		{
-			uint32_t result_type = ops[0];
-			uint32_t id = ops[1];
-			const auto *arg = &ops[3];
-			uint32_t length = instruction.length - 3;
-
-			auto &callee = get<SPIRFunction>(func_id);
-			auto &return_type = get<SPIRType>(callee.return_type);
-			bool pure = function_is_pure(callee);
-
-			bool emit_return_value_as_argument = false;
-
-			for (uint32_t i = 0; i < length; i++)
-			{
-				if (callee.arguments[i].write_count)
-				{
-					register_call_out_argument(arg[i]);
-				}
-
-				flush_variable_declaration(arg[i]);
-			}
-
-			if (!return_type.array.empty() && !backend.can_return_array)
-			{
-				emit_return_value_as_argument = true;
-			}
-
-			if (!pure)
-				register_impure_function_call();
-
-			string funexpr;
-			SmallVector<string> arglist;
-			funexpr += to_name(func_id) + "(";
-
-			if (emit_return_value_as_argument)
-			{
-				statement(type_to_glsl(return_type), " ", to_name(id), type_to_array_glsl(return_type, 0), ";");
-				arglist.push_back(to_name(id));
-			}
-
-			arglist.push_back("meshStream");
-
-			for (uint32_t i = 0; i < length; i++)
-			{
-				if (skip_argument(arg[i]))
-					continue;
-
-				arglist.push_back(to_func_call_arg(callee.arguments[i], arg[i]));
-			}
-
-			begin_scope();
-			for (uint32_t i = length; i < callee.arguments.size(); i++)
-			{
-				uint32_t arg_id = callee.arguments[i].id;
-				// implicit arguments usually alias variables.
-				// If it's not a variable, just pass it (it might be an expression, although rare for implicit args).
-				if (ir.ids[arg_id].get_type() == TypeVariable)
-				{
-					auto &var = get<SPIRVariable>(arg_id);
-					string name = to_name(arg_id);
-					string arg_str = to_func_call_arg(callee.arguments[i], arg_id);
-					auto space = arg_str.find_last_of(' ');
-					if (space != string::npos)
-					{
-						string last_part = arg_str.substr(space + 1);
-						if (last_part == name || (name.find('.') != string::npos && last_part == name.substr(name.find_last_of('.') + 1)))
-							arg_str = last_part;
-					}
-
-					if (name.find('.') == string::npos && arg_str == name)
-					{
-						bool old_deferred = var.deferred_declaration;
-						var.deferred_declaration = false;
-						statement(CompilerGLSL::variable_decl(var), ";");
-						var.deferred_declaration = old_deferred;
-					}
-					arglist.push_back(arg_str);
-				}
-				else
-				{
-					arglist.push_back(to_func_call_arg(callee.arguments[i], arg_id));
-				}
-			}
-
-			for (auto &combined : callee.combined_parameters)
-			{
-				uint32_t image_id = combined.global_image ? combined.global_image : arg[combined.image_id];
-				uint32_t sampler_id = combined.global_sampler ? combined.global_sampler : arg[combined.sampler_id];
-				arglist.push_back(to_combined_image_sampler(image_id, sampler_id));
-			}
-
-			funexpr += merge(arglist);
-			funexpr += ")";
-
-			if (result_type != 0 && get<SPIRType>(result_type).basetype != SPIRType::Void)
-			{
-				// Propagate precision into the function call
-				if (has_decoration(id, DecorationRelaxedPrecision))
-					set_decoration(func_id, DecorationRelaxedPrecision);
-				if (has_decoration(id, DecorationNoContraction))
-					set_decoration(func_id, DecorationNoContraction); // Not actually used by GLSL, but good for tracking.
-
-				emit_op(result_type, id, funexpr, should_forward(func_id));
-			}
-			else
-				statement(funexpr, ";");
-			end_scope();
-		}
-		else
-			CompilerGLSL::emit_instruction(instruction);
-		break;
-	}
 
 	case OpInBoundsAccessChain:
 	case OpAccessChain:
@@ -14042,16 +13926,6 @@
 		}
 	}
 
-	if (!processing_entry_point)
-	{
-		if (msl_options.for_mesh_pipeline && functions_needing_mesh_stream.count(func.self))
-		{
-			if (added_arg)
-				decl += ", ";
-			decl += "thread mesh_stream_t &meshStream";
-			added_arg = true;
-		}
-	}
 
 	if (!func.arguments.empty() && added_arg)
 		decl += ", ";
@@ -14135,9 +14009,22 @@
 		if (&arg != &func.arguments.back())
 			decl += ", ";
 	}
+	if (!processing_entry_point && msl_options.for_mesh_pipeline && functions_needing_mesh_stream.count(func.self))
+	{
+		if (added_arg || !func.arguments.empty())
+			decl += ", ";
+		decl += "thread mesh_stream_t &meshStream";
+	}
 
 	decl += ")";
 	statement(decl);
+}
+
+void CompilerMSL::append_global_func_args(const SPIRFunction &func, uint32_t index, SmallVector<string> &arglist)
+{
+	CompilerGLSL::append_global_func_args(func, index, arglist);
+	if (msl_options.for_mesh_pipeline && functions_needing_mesh_stream.count(func.self))
+		arglist.push_back("meshStream");
 }
 
 static bool needs_chroma_reconstruction(const MSLConstexprSampler *constexpr_sampler)
diff --git a/spirv_msl.hpp b/spirv_msl.hpp
--- a/spirv_msl.hpp
+++ b/spirv_msl.hpp
@@ -989,6 +989,7 @@
 	                                           const uint32_t *args, uint32_t count) override;
 	void emit_header() override;
 	void emit_function_prototype(SPIRFunction &func, const Bitset &return_flags) override;
+	void append_global_func_args(const SPIRFunction &func, uint32_t index, SmallVector<std::string> &arglist) override;
 	void emit_sampled_image_op(uint32_t result_type, uint32_t result_id, uint32_t image_id, uint32_t samp_id) override;
 	void emit_subgroup_op(const Instruction &i) override;
 	void emit_subgroup_cluster_op(uint32_t result_type, uint32_t result_id, uint32_t cluster_size, uint32_t op0,
-- 
2.50.1
